rca/utils/models.py

```diff
@@ -396,6 +398,148 @@ class LegacySiteTaggedPage(ItemBase):
     )
 
 
+class NewsAndEventsMixin:
+    """
+    This class is desiged to first work _with_ LegacyNewsAndEventsMixin and
+    eventually replace it.
+
+    If a page has no ``legacy_news_and_event_tags`` value, then this class can
+    be used to return related editorial and event pages from within this site,
+    rather than via the api. The reasoning for this is that publishing editorial
+    and event items is likely to take time. Adding this logic means that if and
+    when a page in question has enough newly published events end editorial
+    pages related, the ``legacy_news_and_event_tags`` value can be set as blank.
+    This will be instatiated and internal related page data will be returned.
+
+    """
+
+    def __init__(self, page, *args, **kwargs):
+
+        from rca.landingpages.models import LandingPage
+        from rca.events.models import EventDetailPage
+        from rca.editorial.models import EditorialPage
+
+        self.event_page = EventDetailPage
+        self.editorial_page = EditorialPage
+
+        self.page = page
+        self.editorial_items = 3
+        self.page_type = page.__class__.__name__
+        self.page_is_landing_page = issubclass(page.__class__, LandingPage)
+
+    def format_data(self, pages_queryset):
+        """
+        Used to turn queryset data into a list of dicts
+        for the template
+        """
+        items = []
+        for page in pages_queryset:
+            # Quite an elaborate check for the editorial_type to avoid
+            # throwing an error when the page here is an EventDetailPage
+            editorial_type = getattr(page, "editorial_types", None)
+            try:
+                editorial_type = editorial_type.first().type
+            except AttributeError:
+                editorial_type = ""
+
+            PAGE_META_MAPPING = {
+                "EditorialPage": editorial_type,
+                "EventDetailPage": "Event",
+            }
+            meta = PAGE_META_MAPPING.get(page.__class__.__name__, "")
+            editorial_published_date = getattr(page, "published_at", None)
+
+            if editorial_published_date:
+                editorial_published_date = editorial_published_date.strftime(
+                    "%-d %B %Y"
+                )
+
+            PAGE_DESCRIPTION_MAPPING = {
+                "EditorialPage": editorial_published_date,
+                "EventDetailPage": getattr(page, "event_date_short", None),
+            }
+            description = PAGE_DESCRIPTION_MAPPING.get(page.__class__.__name__, "")
```

Comment:
It looks like all of this mapping (and the complication with editorial type above) are the result of this function taking two different page types, and using different fields on each to populate `meta` and `description` fields.

Would it make sense to instead add a `listing_meta` and `listing_description` property to each of these models, and return them in the same way as `page.url` etc?

If not, alternatively you could just do an if, or a single map (with the values being methods) and call eg `format_editorial_page` to return a dictionary, and use that to `.update` the appended item? That way you wouldn't have to consider both page types within this generic function.

Reply:
Yea we've sorted of written ourselves into a corner with this.
It needs to be one generic method as the QS we pass to be formatted is chained and will either be:
[editorial,editorial,event]
or
[editorial,editorial,editorial]

I don't see any blocking issues here rather than readability / preference so, given the time left, I'm going to leave this as is.

Reply:
absolutely - not sure I follow the chaining aspect, but you have clearly given it some thought so happy with that